File: web/js/registration.js
var courses = {};
var sections = {};

var base = "Registration?";

function getDepartments(){
	$("#status")[0].innerHTML = "";
	$("#enroll")[0].value = "Enroll";
	$("#enroll")[0].onclick = function(){ alert("Select a Department!"); };
	var url=base+"dept";
	
	var callback = function(data) {
		var json = parseJSON(data);
		var b = json["results"]["bindings"];
		var html = "<option value='-1'>--</option>";
		for(var i=0; i<b.length; i++){
			var dept = b[i]["department"];
			html += "<option value='"+dept.id+"'>"+dept.name+"</option>\n";
		}
		$("#dept")[0].innerHTML = html;
	};
	
	$.post(url, callback);
}

function getCourses(deptId){
	$("#status")[0].innerHTML = "";
	$("#enroll")[0].value = "Enroll";
	$("#enroll")[0].onclick = function(){ alert("Select a Course!"); };
	if(deptId != -1){
		var url=base+"course="+deptId;
		
		var callback = function(data) {	
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var html = "<option value='-1'>--</option>";
			for(var i=0; i<b.length; i++){
				var course = b[i]["course"];
				courses[course.id] = course.desc;
				html += "<option value='"+course.id+"'>"+course.name+"</option>\n";
			}
			$("#course")[0].innerHTML = html;
			$("#courseInfo")[0].innerHTML = "";
		};
		
		$.post(url, callback);
	}
}

function getSections(courseId){
	$("#status")[0].innerHTML = "";
	$("#enroll")[0].value = "Enroll";
	$("#enroll")[0].onclick = function(){ alert("Select a Section!"); };
	if(courseId != -1){
		$("#courseInfo")[0].innerHTML = courses[courseId];
		var url=base+"section="+courseId;
		
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var html = "<option value='-1'>--</option>";
			for(var i=0; i<b.length; i++){
				var section = b[i]["section"];
				sections[section.id] = section.faculty;
				html += "<option value='"+section.id+"'>"+section.days+"-"+section.period+"</option>\n";
			}
			$("#section")[0].innerHTML = html;
			$("#sectionInfo")[0].innerHTML = "";
		};
		
		$.post(url, callback);
	}
}

function getEnrolled(){
	
	var url=base+"enrolled";
	
	var callback = function(data) {
		var json = parseJSON(data);
		var b = json["results"]["bindings"];
		
		var html = "<table><tbody><tr><th>Course</th><th>Faculty</th><th>Days</th><th>Period</th></tr>";
		for(var i=0; i<b.length; i++){
			var section = b[i]["section"];
			html += "<tr><td>"+section.course+"</td><td>"+section.faculty+"</td><td>"+
			section.days+"</td><td>"+section.period+"</td>"+
			"<td><input type='button' value='Remove' onclick='unEnroll("+section.id+")'/></td></td></tr>\n";
		}
		html +="</table>";
		$("#classes")[0].innerHTML = html;
	};
	
	$.post(url, callback);
}

function isEnrolled(sectionId){
	if(sectionId != -1){
		$("#sectionInfo")[0].innerHTML = "Taught by: " + sections[sectionId];
		var url=base+"selected="+sectionId;
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			if(b[0]["enrolled"] == "true"){
				$("#enroll")[0].value = "Unenroll";
				$("#enroll")[0].onclick = function(){
					var section = $("#section")[0];
					unEnroll(section.options[section.selectedIndex].value);
				};
			}else{
				$("#enroll")[0].value = "Enroll";
				$("#enroll")[0].onclick = function(){
					var section = $("#section")[0];
					enroll(section.options[section.selectedIndex].value);
				};
			}
		};
		$.post(url, callback);
	}
}

function enroll(sectionId){
	if(sectionId != -1){
		var url=base+"enroll="+sectionId;
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var success = b[0]["success"]=="true"?"Enrollment Successful": "Enrollment Unsuccessful";
			$("#status")[0].innerHTML = success;
			$("#enroll")[0].value = "Unenroll";
			$("#enroll")[0].onclick = function(){
				var section = $("#section")[0];
				unEnroll(section.options[section.selectedIndex].value);
			};
			getEnrolled();
		};
		$.post(url, callback);
	}
}

function unEnroll(sectionId){
	if(sectionId != -1){
		var url=base+"unenroll="+sectionId;
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var success = b[0]["success"]=="true"?"Unenrollment Successful": "Unenrollment Unsuccessful";
			$("#status")[0].innerHTML = success;
			$("#enroll")[0].value = "Enroll";
			$("#enroll")[0].onclick = function(){
				var section = $("#section")[0];
				enroll(section.options[section.selectedIndex].value);
			};
			getEnrolled();
		};
		$.post(url, callback);
	}
}

function parseJSON(json){
	return eval("("+json+")");
}

File: src/servlets/Registration.java
import java.io.*;
import java.util.*;

import javax.servlet.*;
import javax.servlet.http.*;

import freemarker.template.*;

import except.*;
import model.ctrl.*;
import model.entity.*;

public class Registration extends HttpServlet {
  static  String templateDir = "WEB-INF/templates";
  private Configuration  cfg; 

  public void init() {
    // Prepare the FreeMarker configuration;
    // - Load templates from the WEB-INF/templates directory of the Web app.
    //
    cfg = new Configuration();
    cfg.setServletContextForTemplateLoading( getServletContext(), "WEB-INF/templates");
  }
  
  public void doGet( HttpServletRequest  req, HttpServletResponse res ) throws ServletException, IOException {
 	Template resultTemplate = null;
    BufferedWriter out = null;
    User u = null;
    String resultTemplateName = "Registration.ftl",
    		error = null;

    HttpSession session = req.getSession();
	u = (User) session.getAttribute( "user" );
    
    // Setup the data-model
    //
    Map vars = new HashMap();

    // Build the data-model
    //
    if(u != null){
    	
    	
    }else resultTemplateName = "Login.ftl";


    // Load templates from the WEB-INF/templates directory of the Web app.
    //
    try {
      resultTemplate = cfg.getTemplate( resultTemplateName );
    }catch (IOException e) {
      throw new ServletException("Can't load template in: " + templateDir + ": " + e.toString());
    }

    // Prepare the HTTP response:
    // - Use the charset of template for the output
    // - Use text/html MIME-type
    //
    out = new BufferedWriter(new OutputStreamWriter( res.getOutputStream(), resultTemplate.getEncoding() ));
    
    
    res.setContentType("text/html; charset=" + resultTemplate.getEncoding());

    // Merge the data-model and the template
    //
    try {
      resultTemplate.process( vars, out );
      out.flush();
    }catch (TemplateException e) {
      throw new ServletException( "Error while processing template", e);
    }
    out.close();
  }

  public void doPost( HttpServletRequest  req, HttpServletResponse res ) throws ServletException, IOException {
  	PrintWriter out = res.getWriter();
    res.setContentType("text/html; charset=UTF-8");
    
    String dept = req.getParameter( "dept" );
    String course = req.getParameter( "course" );
    String section = req.getParameter( "section" );
    String selected = req.getParameter( "selected" );
    String enroll = req.getParameter( "enroll" );
    String unenroll = req.getParameter( "unenroll" );
    String enrolled = req.getParameter( "enrolled" );
    
    HttpSession session = req.getSession();
	User u = (User) session.getAttribute( "user" );
    
    StringBuffer content = new StringBuffer();
    
	content.append("{\n");	content.append("  \"results\": {\n");	content.append("    \"bindings\": [\n\n");    
    
    if(dept != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getDepartments(); iter.hasNext();){
	    		Department d = (Department) iter.next();
	    		content.append("      {\n");				content.append("        \"department\": { \"id\": \""+d.getId()+"\", \"name\": \""+d.getName()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
    	}catch(PortsException e){}
    }else if(course != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getCourses(Integer.parseInt(course)); iter.hasNext();){
	    		Course c = (Course) iter.next();
	    		content.append("      {\n");				content.append("        \"course\": { \"id\": \""+c.getId()+"\", \"dept\": \""+
											c.getDeptId()+"\", \"name\": \""+c.getName()+"\", \"desc\": \""+c.getDesc()+"\" }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }else if(section != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getSections(Integer.parseInt(section)); iter.hasNext();){
	    		Section s = (Section) iter.next();
	    		Faculty f = CtrlRegistration.getFaculty(s.getFacultyId());
	    		String facName = f.getName();
	    		content.append("      {\n");				content.append("        \"section\": { \"id\": \""+s.getId()+"\", \"course\":\""+
										s.getCourseId()+"\", \"faculty\": \""+facName+"\", \"days\": \""+
										s.getDays()+"\", \"period\": \""+s.getPeriod()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }else if(selected != null){
    	content.append("      {\n");
    	try{
	    	int sectionId = Integer.parseInt(selected);
	    	if(CtrlRegistration.studentOnRoll(sectionId, u.getId()))
	    		content.append("        \"enrolled\": \"true\"\n");
	    	else content.append("        \"enrolled\": \"false\"\n");
	    }catch(PortsException e){}
	    content.append("      } ,\n");
    }else if(enroll != null){
    	content.append("      {\n");
    	try{
	    	int sectionId = Integer.parseInt(enroll);
	    	CtrlRegistration.enrollStudent(sectionId, u.getId());
	    	content.append("        \"success\": \"true\"\n");
	    }catch(PortsException e){content.append("        \"success\": \"false\"\n");}
	    content.append("      } ,\n");
    }else if(unenroll != null){
    	content.append("      {\n");
    	try{
	    	int sectionId = Integer.parseInt(unenroll);
	    	CtrlRegistration.unEnrollStudent(sectionId, u.getId());
	    	content.append("        \"success\": \"true\"\n");
	    }catch(PortsException e){content.append("        \"success\": \"false\"\n");}
	    content.append("      } ,\n");
    }else if(enrolled != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.studentSections(u.getId()); iter.hasNext();){
	    		Section s = (Section) iter.next();
	    		Faculty f = CtrlRegistration.getFaculty(s.getFacultyId());
	    		Course c = CtrlRegistration.getCourse(s.getCourseId());
	    		String facName = f.getName();
	    		content.append("      {\n");				content.append("        \"section\": { \"id\": \""+s.getId()+"\", \"course\":\""+
										c.getName()+"\", \"faculty\": \""+facName+"\", \"days\": \""+
										s.getDays()+"\", \"period\": \""+s.getPeriod()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }
    content.append("    ]\n");	content.append("  }\n");	content.append("}\n");
    out.println(content.toString());
  }

}

